File: BingoVerifier.java
/*
 * Xavier Jibril Goudeaux
 * Java Bingo Verifier
 * Due:9/11/2022
 * 
 */

import java.util.Scanner;

public class BingoVerifier {

	// Rotates player card 90 degrees to the right
	public void rotateCard90(String[][] card) {
		int n = card.length;
		for (int row = 0; row < n / 2; row++) {
			for (int column = row; column < n - row - 1; column++) {
				String temp = card[row][column];
				card[row][column] = card[n - 1 - column][row];
				card[n - 1 - column][row] = card[n - 1 - row][n - 1 - column];
				card[n - 1 - row][n - 1 - column] = card[column][n - 1 - row];
				card[column][n - 1 - row] = temp;
			}
		}
	}

	/*
	 * Indicates 'straight' pattern (return 1) or 'crazy' pattern (return 4)
	 * Return false if neither patterns are found
	 */
	public int patternType(String[][] pattern) {
		for (int row = 0; row < pattern.length; row++) {
			for (int column = 0; column < pattern[row].length; column++) {
				if (pattern[row][column].equals("1")) {
					return 1;
				}
				if (pattern[row][column].equals("4")) {
					return 4;
				}
			}
		}
		return -1;
	}

	// Marks positions of called numbers on player card. Includes free space '00'
	public void markCard(String[] calledNumbers, String[][] playerCard) {
		for (String num : calledNumbers) {
			for (int row = 0; row < playerCard.length; row++) {
				for (int column = 0; column < playerCard[row].length; column++) {
					if (num.equals(playerCard[row][column]) || playerCard[row][column] == "00") {
						playerCard[row][column] = "X" + playerCard[row][column];
					}
				}
			}
		}
	}

	/*
	 * Checks if marked positions are the same as straight or crazy pattern.
	 * Returns false if marked card positions don't match input pattern
	 */
	public boolean winPattern(String[][] pattern, String[][] playerCard) {
		for (int row = 0; row < playerCard.length; row++) {
			for (int column = 0; column < playerCard[row].length; column++) {
				if ((pattern[row][column].equals("1") || pattern[row][column].equals("4"))
						&& playerCard[row][column].charAt(0) != 'X') {
					return false;
				}
			}
		}
		return true;
	}

	/*
	 * Checks to see if the last number called is a part of bingo pattern.
	 * Returns true if last number called is in the pattern
	 */
	public boolean lastCalled(String[] calledNumbers, String[][] playerCard, String pattern[][]) {
		String last = calledNumbers[calledNumbers.length - 1];
		for (int row = 0; row < pattern.length; row++) {
			for (int column = 0; column < pattern[row].length; column++) {
				if (playerCard[row][column].substring(0, 1).equals("X")
						&& playerCard[row][column].substring(1).equals(last)) {
					if (pattern[row][column].equals("1") || pattern[row][column].equals("4")) {
						return true;
					}
				}
			}
		}
		return false;
	}

	// Verify player card against called numbers and pattern card for bingo
	/*
	 * 1. Mark off player card based on numbers called out
	 * 2. Indicate 'Straight' Pattern or 'Crazy' Pattern
	 * 3a. check if winning pattern has been marked off AND that the last number
	 * called is in that pattern
	 * 3b. If 'Crazy', do step 3a then rotate card 90 degrees. Repeat until
	 * conditions are met or all of the rotations fail
	 * 4. Return true if 3a condition is met.
	 */
	public boolean verifyBingo(String[] calledNumbers, String[][] playerCard, String[][] pattern) {
		markCard(calledNumbers, playerCard);
		if (patternType(pattern) == 1) {
			if (lastCalled(calledNumbers, playerCard, pattern) && winPattern(pattern, playerCard)) {
				return true;
			}
		}
		if (patternType(pattern) == 4) {
			for (int i = 0; i < 4; i++) {
				if (lastCalled(calledNumbers, playerCard, pattern) && winPattern(pattern, playerCard)) {
					return true;
				}
				rotateCard90(playerCard);
			}
		}
		return false;
	}

	// displays pattern card or player card to terminal
	public void displayInput(String[][] card) {
		for (int row = 0; row < card.length; row++) {
			for (int column = 0; column < card[row].length; column++) {
				System.out.print(card[row][column] + " ");
			}
			System.out.println();
		}
		System.out.println();
	}

	// displays called numbers to terminal
	public void displayInput(String[] calledNumbers) {
		for (int index = 0; index < calledNumbers.length; index++) {
			System.out.print(calledNumbers[index] + " ");
		}
		System.out.println();
		System.out.println();
	}

	public static void main(String[] args) {
		// Create scanner for input
		Scanner scan = new Scanner(System.in);

		// Create BingoVerifier Object to use functions
		BingoVerifier bingo = new BingoVerifier();

		String[][] pattern;
		String numbers;
		String[] calledNumbers;
		String[][] playerCard;

		System.out.println("------------------------------");
		System.out.println("INPUT:");
		// Pattern
		pattern = new String[5][5];
		for (int row = 0; row < pattern.length; row++) {
			for (int column = 0; column < pattern[row].length; column++) {
				pattern[row][column] = scan.next();
			}
		}

		/*
		 * Consume blank line
		 */
		scan.nextLine();
		scan.nextLine();

		// Called Numbers
		numbers = scan.nextLine();
		calledNumbers = numbers.split(" ");

		/*
		 * Consume blank line
		 */
		scan.nextLine();

		// Player Card
		playerCard = new String[5][5];
		for (int row = 0; row < playerCard.length; row++) {
			for (int column = 0; column < playerCard[row].length; column++) {
				playerCard[row][column] = scan.next();
			}
		}
		// close
		scan.close();

		/*
		 * Performs bingo verification using inputs pattern, called numbers, and player
		 * card.
		 * Outputs whether or not player has bingo.
		 */

		System.out.println("\nOUTPUT:");
		if (bingo.verifyBingo(calledNumbers, playerCard, pattern)) {
			System.out.println("VALID BINGO\n\r");
		} else {
			System.out.println("NO BINGO\n\r");
		}
	}
}

File: BingoVerifier.py
#Xavier Goudeaux
#Python Bingo Verifier
#10/18/2022


def rotate_90(card):
    n = len(card)
    for row in range(int(n/2)):
        for col in range(row,n - row - 1):
            temp = card[row][col]
            card[row][col] = card[n - 1 - col][row]
            card[n - 1 - col][row] = card[n - 1 - row][n - 1 - col]
            card[n - 1 - row][n - 1 - col] = card[col][n - 1 - row]
            card[col][n - 1 - row] = temp

def pattern_type(pattern):
    for row in pattern:
        if '1' in row:
            return 1
        elif '4' in row:
            return 4
    return -1
        
        		
def win_pattern(pattern, card):
    for row in range(5):
        for col in range(len(pattern[row])):
            if (pattern[row][col] == '1' or pattern[row][col] == '4') and card[row][col][0] != 'X':
                return False
    return True

#last called 
def last_called(pattern, numbers, card):
    last = numbers[-1]
    for row in range(len(card)):
        for col in range(len(card[row])):
            if (card[row][col][0] == 'X' and card[row][col][1:] == last) and (pattern[row][col] == '1' or pattern[row][col] == '4'):
                return True
    return False

#verifies if player has a valid bingo
def verify_bingo(pattern,numbers, card):
    mark_card(numbers, card)
    if pattern_type(pattern) == 1:
        if last_called(pattern, numbers, card) and win_pattern(pattern, card):
            return True
    if pattern_type(pattern) == 4:
        for rotations in range(4):
            if last_called(pattern, numbers, card) and win_pattern(pattern, card):
                return True
            rotate_90(card)
    return False

#marks off called numbers from player card with 'X' symbol
def mark_card(numbers,card):
    for num in numbers:
        for row in range(len(card)):
            for col in range(len(card[row])):
                if card[row][col] == num or card[row][col] == "00":
                    card[row][col] = "X" + card[row][col]

#displays 1d list inputs
def display_1d(numbers):
    for num in numbers:
        print(num, end=' ')
    print()

#displays 2d list inputs
def display_2d(card):
    for row in card:
        for item in row:
            print(item, end=' ')
        print()
    print()

def main():
    print('-'*30)
    print('INPUT:')
    
    inputs = [input() for _ in range(13)]
    input_list = [row for row in inputs if row != '']
    pattern = [input_list[row].split() for row in range(5) ]
    numbers = input_list[5].split()
    card = [input_list[row].split() for row in range(6,len(input_list))]

    print('\nOUTPUT:')
    if(verify_bingo(pattern,numbers, card)):
        print('VALID BINGO\n')
    else:
        print('NO BINGO\n')

if __name__ == "__main__":
    main()